core/clingutils/src/TClingUtils.cxx

```diff
@@ -5279,6 +5279,32 @@ int ROOT::TMetaUtils::AST2SourceTools::FwdDeclFromTmplDecl(const clang::Template
    return EncloseInNamespaces(templDecl, defString);
 }
 
+////////////////////////////////////////////////////////////////////////////////
+/// Convert a tmplt decl to its fwd decl
+
+int ROOT::TMetaUtils::AST2SourceTools::FwdDeclIfTmplSpec(const clang::RecordDecl& recordDecl,
+                                                         const cling::Interpreter& interpreter,
+                                                         std::string& defString,
+                                                         const std::string &normalizedName)
+{
+   // If this is an explicit specialization, inject it into cling, too, such that it can have
+   // externalLexicalStorage, see TCling.cxx's ExtVisibleStorageAdder::VisitClassTemplateSpecializationDecl.
+   if (auto tmplSpecDeclPtr = llvm::dyn_cast<clang::ClassTemplateSpecializationDecl>(&recordDecl)) {
+      if (const auto *specDef = tmplSpecDeclPtr->getDefinition()) {
+         if (specDef->getTemplateSpecializationKind() == clang::TSK_ExplicitSpecialization) {
+            // normalizedName contains scope, no need to enclose in namespace!
+            defString += "template <> class " + normalizedName + ';';
```

Comment:
@Axel-Naumann, it seems that patch is not sufficient (Adjusted later by 7a8a81225d05c0debe2882b9bb48db060b53e1c9). On the cmssw side we get an error which seem related:

```[vvasilev@cmsdev31 CMSSW_11_3_CXXMODULE_X_2021-01-29-2300-cms]$ root.exe -l -b -q
Forward declarations from ./CUDADataFormatsCaloCommon_xr.rootmap:6:30: error: explicit instantiation of undefined template 'calo::common::AddSize<calo::common::tags::Ptr>'
template class calo::common::AddSize<calo::common::tags::Ptr>;
                             ^
Forward declarations from ./CUDADataFormatsCaloCommon_xr.rootmap:4:67: note: template is declared here
namespace calo { namespace common { template <typename tag> class AddSize; } }
                                                                  ^
Forward declarations from ./CUDADataFormatsCaloCommon_xr.rootmap:8:30: error: explicit instantiation of undefined template 'calo::common::AddSize<calo::common::tags::DevPtr>'
template class calo::common::AddSize<calo::common::tags::DevPtr>;
                             ^
Forward declarations from ./CUDADataFormatsCaloCommon_xr.rootmap:4:67: note: template is declared here
namespace calo { namespace common { template <typename tag> class AddSize; } }
                                                                  ^
root.exe: /build/dlange/210130/BUILD/slc7_amd64_gcc900/lcg/root/6.23.01/root-6.23.01/interpreter/cling/lib/Interpreter/Transaction.cpp:136: void cling::Transaction::append(cling::Transaction::DelayCallInfo): Assertion `getState() == kCollecting && "Cannot append declarations in current state."' failed.
```

cc: @davidlange6

Reply:
Thanks - can I see the `genreflex` invocation, the umbrella header being passed (`classes_xr.h` or whatever it's called) and `selection.xml`?

Reply:
Btw this looks like what you tried to fix (but only for `std::`) in https://github.com/root-project/root/pull/6385/commits/7a8a81225d05c0debe2882b9bb48db060b53e1c9 - once this CMSSW issue is understood we might want to review / revert https://github.com/root-project/root/pull/6385/commits/7a8a81225d05c0debe2882b9bb48db060b53e1c9.

Reply:
classes_def.xml

```
<lcgdict>
    <enum name="calo::common::tags::Ptr"/>
    <enum name="calo::common::tags::Vec"/>
    <enum name="calo::common::tags::DevPtr"/>
    <class name="calo::common::AddSize<calo::common::tags::Ptr>"/>
    <class name="calo::common::AddSize<calo::common::tags::Vec>"/>
    <class name="calo::common::AddSize<calo::common::tags::DevPtr>"/>
</lcgdict>
```
and classes.h
```
#include "DataFormats/Common/interface/Wrapper.h"
#include "CUDADataFormats/CaloCommon/interface/Common.h"
```
and the genreflex (apologies for the long random paths..)
```/build/dlange/210130/slc7_amd64_gcc900/lcg/root/6.23.01/bin/genreflex src/CUDADataFormats/CaloCommon/src/classes.h -s src/CUDADataFormats/CaloCommon/src/classes_def.xml -o tmp/slc7_amd64_gcc900/src/CUDADataFormats/CaloCommon/src/CUDADataFormatsCaloCommon/a/CUDADataFormatsCaloCommon_xr.cc --fail_on_warnings --rootmap=tmp/slc7_amd64_gcc900/src/CUDADataFormats/CaloCommon/src/CUDADataFormatsCaloCommon/a/CUDADataFormatsCaloCommon_xr.rootmap --rootmap-lib=libCUDADataFormatsCaloCommon.so -DCMS_DICT_IMPL -D_REENTRANT -DGNUSOURCE -D__STRICT_ANSI__ -DGNU_GCC -D_GNU_SOURCE -DEIGEN_DONT_PARALLELIZE -DTBB_USE_GLIBCXX_VERSION=90300 -DTBB_SUPPRESS_DEPRECATED_MESSAGES -DBOOST_SPIRIT_THREADSAFE -DPHOENIX_THREADSAFE -DBOOST_MATH_DISABLE_STD_FPCLASSIFY -DBOOST_UUID_RANDOM_PROVIDER_FORCE_POSIX -DCMSSW_GIT_HASH="CMSSW_11_3_CXXMODULE_X_2021-01-29-2300" -DPROJECT_NAME="CMSSW" -DPROJECT_VERSION="CMSSW_11_3_CXXMODULE_X_2021-01-29-2300-cms" -I/build/dlange/210130/tmp/BUILDROOT/bc659db780b87998cdd721195aad8098/opt/cmssw/slc7_amd64_gcc900/cms/cmssw/CMSSW_11_3_CXXMODULE_X_2021-01-29-2300-cms/src -I/build/dlange/210130/slc7_amd64_gcc900/external/pcre/8.43-bcolbf/include -isystem/build/dlange/210130/slc7_amd64_gcc900/external/boost/1.72.0-727ddbde87b1a029a07e1d0cf14b886c/include -I/build/dlange/210130/slc7_amd64_gcc900/external/bz2lib/1.0.6-bcolbf2/include -I/build/dlange/210130/slc7_amd64_gcc900/external/cuda/11.2.0/include -I/build/dlange/210130/slc7_amd64_gcc900/external/libuuid/2.34-bcolbf2/include -isystem/build/dlange/210130/slc7_amd64_gcc900/lcg/root/6.23.01/include -isystem/build/dlange/210130/slc7_amd64_gcc900/external/tbb/2020_U3/include -I/build/dlange/210130/slc7_amd64_gcc900/external/xz/5.2.4-bcolbf2/include -I/build/dlange/210130/slc7_amd64_gcc900/external/zlib/1.2.11-bcolbf/include -I/build/dlange/210130/slc7_amd64_gcc900/external/eigen/d812f411c3f9-ghbfee/include/eigen3 -I/build/dlange/210130/slc7_amd64_gcc900/external/fmt/7.0.1/include -I/build/dlange/210130/slc7_amd64_gcc900/external/md5/1.0.0-bcolbf2/include -I/build/dlange/210130/slc7_amd64_gcc900/external/tinyxml2/6.2.0-ghbfee/include -DCMSSW_REFLEX_DICT
```

Reply:
Thanks! Last bit, hopefully: where is `calo::common::AddSize` defined?

Reply:
@davidlange6 @vgvassilev what's the current version of this error message? Current as of https://github.com/root-project/root/pull/6385/commits/be64f2bb74ea893f69c6350c926863a7e2348ca4 that I merged just a couple of seconds ago?